lightning-persister/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+		let filename = self.get_full_filepath(funding_txo);
+		let tmp_filename = format!("{}.tmp", filename.clone());
+
+		{
+			// Note that going by rust-lang/rust@d602a6b, on MacOS it is only safe to use
+			// rust stdlib 1.36 or higher.
+			let mut f = fs::File::create(&tmp_filename)?;
+			monitor.write(&mut f)?;
+			f.sync_all()?;
+		}
+		fs::rename(&tmp_filename, &filename)?;
+		// Fsync the parent directory on Unix.
+		#[cfg(not(target_os = "windows"))]
+		{
+			let path = Path::new(&filename).parent().unwrap();
+			let dir_file = fs::OpenOptions::new().read(true).open(path)?;
+			unsafe { libc::fsync(dir_file.as_raw_fd()); }
+		}
+		Ok(())
+	}
+
+	#[cfg(test)]
+	fn load_channel_data<ChanSigner: ChannelKeys + Readable + Writeable>(&self) ->
+		Result<HashMap<OutPoint, ChannelMonitor<ChanSigner>>, ChannelMonitorUpdateErr> {
+		if let Err(_) = fs::create_dir_all(&self.path_to_channel_data) {
+			return Err(ChannelMonitorUpdateErr::PermanentFailure);
+		}
+		let mut res = HashMap::new();
+		for file_option in fs::read_dir(&self.path_to_channel_data).unwrap() {
+			let file = file_option.unwrap();
+			let owned_file_name = file.file_name();
+			let filename = owned_file_name.to_str();
+			if !filename.is_some() || !filename.unwrap().is_ascii() || filename.unwrap().len() < 65 {
+				return Err(ChannelMonitorUpdateErr::PermanentFailure);
+			}
+
+			let txid = Txid::from_hex(filename.unwrap().split_at(64).0);
+			if txid.is_err() { return Err(ChannelMonitorUpdateErr::PermanentFailure); }
+
+			let index = filename.unwrap().split_at(65).1.split('.').next().unwrap().parse();
+			if index.is_err() { return Err(ChannelMonitorUpdateErr::PermanentFailure); }
+
+			let contents = fs::read(&file.path());
+			if contents.is_err() { return Err(ChannelMonitorUpdateErr::PermanentFailure); }
+
+			if let Ok((_, loaded_monitor)) =
+				<(BlockHash, ChannelMonitor<ChanSigner>)>::read(&mut Cursor::new(&contents.unwrap())) {
+				res.insert(OutPoint { txid: txid.unwrap(), index: index.unwrap() }, loaded_monitor);
+			} else {
+				return Err(ChannelMonitorUpdateErr::PermanentFailure);
+			}
+		}
+		Ok(res)
+	}
+}
+
+impl<ChanSigner: ChannelKeys + Readable + Writeable + Send + Sync> channelmonitor::Persist<ChanSigner> for FilesystemPersister {
+	fn persist_new_channel(&self, funding_txo: OutPoint, monitor: &ChannelMonitor<ChanSigner>) -> Result<(), ChannelMonitorUpdateErr> {
+		self.write_channel_data(funding_txo, monitor)
+		  .map_err(|_| ChannelMonitorUpdateErr::PermanentFailure)
+	}
+
+	fn update_persisted_channel(&self, funding_txo: OutPoint, _update: &ChannelMonitorUpdate, monitor: &ChannelMonitor<ChanSigner>) -> Result<(), ChannelMonitorUpdateErr> {
+		self.write_channel_data(funding_txo, monitor)
+		  .map_err(|_| ChannelMonitorUpdateErr::PermanentFailure)
+	}
+}
+
+impl Drop for FilesystemPersister {
+	fn drop(&mut self) {
+		fs::remove_dir_all(&self.path_to_channel_data).unwrap();
+	}
+}
```

Comment:
I think we'll need to configure this only for testing (or possibly moving it into the `tests ` module?). Otherwise, anything persisted would be deleted after the persister is dropped in actual use.

Reply:
Really good catch